### ISSUE TITLE
Hibernate ORM and Hibernate Reactive cannot be used in the same application

### ISSUE THREAD

**Issue by xuxianzhang** [quarkusio/quarkus#13425]
**Describe the bug**
when i boot both **hibernate-reactive and hibernate-orm** in my own project it throw an **IllegalStateException** and Quarkus startup failure

My application.yaml:
```yaml
quarkus:
  datasource:
    jdbc:
      url: ${JDBC_DATABASE_URL:jdbc:mysql://mysql:3306/temple}
    reactive:
      url: ${DATABASE_URL:mysql://mysql:3306/temple}
```

I think it's probably because a private method **checkIsReactive** in **PreconfiguredReactiveServiceRegistryBuilder.class** .
This inspection is not supposed to be here.
It should  be put in **FastBootHibernateReactivePersistenceProvider.getEntityManagerFactoryBuilderOrNull** where the RecordedState be created at.
```java
           RecordedState recordedState = PersistenceUnitsHolder.getRecordedState(persistenceUnitName);
           if(!recordedState.isReactive()){
                log.debug("An Hibernate Reactive serviceregistry can't boot on a non-reactive RecordedState!");
                continue;
           }


            final PrevalidatedQuarkusMetadata metadata = recordedState.getMetadata();
            final BuildTimeSettings buildTimeSettings = recordedState.getBuildTimeSettings();
            final IntegrationSettings integrationSettings = recordedState.getIntegrationSettings();
            RuntimeSettings.Builder runtimeSettingsBuilder = new RuntimeSettings.Builder(buildTimeSettings,
                    integrationSettings);
```

This is more in keeping with the original purpose of the method **getEntityManagerFactoryBuilderOrNull**

**Comment by gsmet** [comment]
Using both Hibernate ORM and Hibernate Reactive in parallel is not supported at the moment. That's something we need to work on.

**Comment by JWood48** [comment]
Any work on this?

**Comment by MartinX3** [comment]
This issue is blocking my workaround for https://github.com/quarkusio/quarkus/issues/19302

**Comment by kenyee** [comment]
Got bit trying to mix these too:
```
    implementation("io.quarkus:quarkus-hibernate-reactive")
    implementation("io.quarkus:quarkus-hibernate-reactive-panache")
    implementation("io.quarkus:quarkus-hibernate-orm-panache-kotlin")
    implementation("io.quarkus:quarkus-jdbc-mysql")
    implementation("io.quarkus:quarkus-reactive-mysql-client")
```
You can't use the hibernate reactive artifacts with panache-kotlin?

**Comment by mklueh** [comment]
I'm running into the same issue after trying to migrate to reactive dependencies. I do not know why this issue occurs. Not using Kotlin.


```
        //implementation 'io.quarkus:quarkus-agroal'
        //implementation 'io.quarkus:quarkus-hibernate-orm'
        //implementation 'io.quarkus:quarkus-hibernate-orm-panache'
        //implementation 'io.quarkus:quarkus-jdbc-postgresql'

        implementation 'io.quarkus:quarkus-reactive-pg-client'
        implementation "io.quarkus:quarkus-hibernate-envers"
        implementation 'io.quarkus:quarkus-hibernate-reactive'
        implementation 'io.quarkus:quarkus-hibernate-validator'
        implementation 'io.quarkus:quarkus-hibernate-reactive-panache'

```

**Comment by kenyee** [comment]
I'd guess that quarkus-hibernate-envers is not reactive @mklueh.  Have you tried removing it?

**Comment by mklueh** [comment]
@kenyee yes I did, nothing changed. 
I haven't tried the hibernate validation dependency, but I'd guess those basic things should work with reactive as well

**Comment by kenyee** [comment]
you could try dumping dependencies... e.g. "./gradlew dependencies" to see what pulls in the non-reactive hibernate...

**Comment by mklueh** [comment]
@kenyee thanks, I'll try that. Does it have to be a non-reactive Hibernate dependency or could it also be something else?

**Comment by kenyee** [comment]
@mklueh mixing isn't allowed...so remove the thing that's pulling in the non-reactive version...

**Comment by mklueh** [comment]
@kenyee This is the only thing I've found
```
+--- io.quarkus:quarkus-hibernate-reactive -> 2.6.3.Final
|    +--- io.quarkus:quarkus-hibernate-orm:2.6.3.Final
```

hibernate-orm is a transitive dependency of hibernate reactive

**Comment by kenyee** [comment]
sounds like you're ok then...can you post your full stack trace?

**Comment by mklueh** [comment]
I've excluded the validator dependency too test-wise, without any effect. This is the full stacktrace:

```
2022-01-25 21:41:51,042 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:122)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:456)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: javax.persistence.PersistenceException: Unable to build EntityManagerFactory
	at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:72)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass.startAll$$superforward1(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass$$function$$5.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:51)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass.startAll(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:97)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy(Unknown Source)
	... 13 more
Caused by: javax.persistence.PersistenceException: Unable to build EntityManagerFactory
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.createEntityManagerFactory(FastBootHibernateReactivePersistenceProvider.java:93)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:149)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:58)
	... 1 more
Caused by: java.lang.IllegalStateException: Booting an Hibernate Reactive serviceregistry on a non-reactive RecordedState!
	at io.quarkus.hibernate.reactive.runtime.boot.registry.PreconfiguredReactiveServiceRegistryBuilder.checkIsReactive(PreconfiguredReactiveServiceRegistryBuilder.java:77)
	at io.quarkus.hibernate.reactive.runtime.boot.registry.PreconfiguredReactiveServiceRegistryBuilder.<init>(PreconfiguredReactiveServiceRegistryBuilder.java:67)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.rewireMetadataAndExtractServiceRegistry(FastBootHibernateReactivePersistenceProvider.java:177)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.getEntityManagerFactoryBuilderOrNull(FastBootHibernateReactivePersistenceProvider.java:156)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.createEntityManagerFactory(FastBootHibernateReactivePersistenceProvider.java:82)
	... 5 more

2022-01-25 21:41:51,042 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:122)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:456)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: javax.persistence.PersistenceException: Unable to build EntityManagerFactory
	at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:72)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass.startAll$$superforward1(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass$$function$$5.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:51)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.quarkus.hibernate.orm.runtime.JPAConfig_Subclass.startAll(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:97)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy(Unknown Source)
	... 13 more
Caused by: javax.persistence.PersistenceException: Unable to build EntityManagerFactory
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.createEntityManagerFactory(FastBootHibernateReactivePersistenceProvider.java:93)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:149)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:58)
	... 1 more
Caused by: java.lang.IllegalStateException: Booting an Hibernate Reactive serviceregistry on a non-reactive RecordedState!
	at io.quarkus.hibernate.reactive.runtime.boot.registry.PreconfiguredReactiveServiceRegistryBuilder.checkIsReactive(PreconfiguredReactiveServiceRegistryBuilder.java:77)
	at io.quarkus.hibernate.reactive.runtime.boot.registry.PreconfiguredReactiveServiceRegistryBuilder.<init>(PreconfiguredReactiveServiceRegistryBuilder.java:67)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.rewireMetadataAndExtractServiceRegistry(FastBootHibernateReactivePersistenceProvider.java:177)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.getEntityManagerFactoryBuilderOrNull(FastBootHibernateReactivePersistenceProvider.java:156)
	at io.quarkus.hibernate.reactive.runtime.FastBootHibernateReactivePersistenceProvider.createEntityManagerFactory(FastBootHibernateReactivePersistenceProvider.java:82)
	... 5 more

```

**Comment by kenyee** [comment]
hmm...this still makes it look like you're mixing them:
"Caused by: java.lang.IllegalStateException: Booting an Hibernate Reactive serviceregistry on a non-reactive RecordedState!
	at io.quarkus.hibernate.reactive.runtime.boot.registry.PreconfiguredReactiveServiceRegistryBuilder.checkIsReactive(PreconfiguredReactiveServiceRegistryBuilder.java:77)"

Can you look at the dependencies on all the other modules?  Though running dependencies at the top should check all the others too 🤔

**Comment by mklueh** [comment]
@kenyee I ran it on the application level, which includes all sub modules and I've searched through the whole dependency tree. 

The hibernate dependencies are included in the root build.gradle and not in the modules directly, so I don't have to change every build.gradle.

Is there a chance to see the specific cause I'm the logs when changing the log level?

**Comment by Sanne** [comment]
If you could share a reproducer I'd like to have a look.  To be clear, mixing is still not expected to work, but it looks like we should at very least improve the error messages or the diagnosticts to figure out what's happening in this case.

**Comment by mklueh** [comment]
It turned out I had removed the Hibernate dependencies, but kept the JDBC dependency

```
implementation 'io.quarkus:quarkus-jdbc-postgresql'
```


I was also running into this

```
Caused by: io.quarkus.runtime.configuration.ConfigurationException: Unable to find a JDBC driver corresponding to the database kind 'postgresql' for the default datasource. Either provide a suitable JDBC driver extension, define the driver manually, or disable the JDBC datasource by adding 'quarkus.datasource.jdbc=false' to your configuration if you don't need it.
	at io.quarkus.agroal.deployment.AgroalProcessor.resolveDriver(AgroalProcessor.java:331)
	at io.quarkus.agroal.deployment.AgroalProcessor.getAggregatedConfigBuildItems(AgroalProcessor.java:277)
	at io.quarkus.agroal.deployment.AgroalProcessor.build(AgroalProcessor.java:82)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

as Quarkus seems to be still sensitive to the non-reactive properties and I had the db kind set `quarkus.datasource.db-kind=postgresql`.

I didn't know that the jdbc dependencies would also not be allowed together with Hibernate Reactive.

Thanks for your help.

**Comment by kenyee** [comment]
@mklueh did you have to remove that property to fix that error?
I tried converting to reactive-hibernate and reverted because I hit that error and couldn't find out why Agroal was complaining about it despite not having a non-reactive JDBC driver...

**Comment by mklueh** [comment]
@kenyee yes, this helps, but also setting `quarkus.datasource.jdbc=false`.

This means (pls correct me if I'm wrong), when you can't use JDBC together with reactive libraries, you cannot use the EntityManager, as it relies on JDBC. Meaning I have to migrate my EntityManager-based queries now to something else, that is provided by reactive libraries, ex.: `io.quarkus:quarkus-reactive-pg-client`

**Comment by kenyee** [comment]
> one could migrate gradually to reactive, and leave some endpoints non-reactive

I don't think you can do that w/o creating two microservices....the JDBC and reactive libraries seem to conflict.  That means no EntityManager as well...

**Comment by yesvike** [comment]
@kenyee If I understand this right, I cannot use a hibernate reactive datasource and a non-reactive datasource in the same app right ?

**Comment by kenyee** [comment]
@yesvike yes...see 2nd comment above.  For your case, that other bug is blocking your usage of db2-reactive so you'll need another service...

**Comment by yesvike** [comment]
> @yesvike yes...see 2nd comment above. For your case, that other bug is blocking your usage of db2-reactive so you'll need another service...

@kenyee before I give up this just wanted to make sure that a hibernate orm reactive data source and plain jdbc data source (no hibernate orm also here) cannot exist together right. I just want to make sure that you did not think the second in non reactive but still hibernate orm. Thanks.

**Comment by kenyee** [comment]
@yesvike: What's interesting is if you dump hibernate-reactive's dependencies, it includes the plain JDBC one.
So if you're going to do raw JDBC using e.g. Jooby or something, then I think you'll be ok....

**Comment by yesvike** [comment]
> @yesvike: What's interesting is if you dump hibernate-reactive's dependencies, it includes the plain JDBC one. So if you're going to do raw JDBC using e.g. Jooby or something, then I think you'll be ok....

Thanks Kenyee for your responses. I cannot dump the hibernate-reactive dependencies in the existing code. Please elaborate on second line. Are you suggesting to use any other Java datasource connection pool mechanism like Hikari ?

**Comment by kenyee** [comment]
./gradlew dependencies
will show you the hierarchy of dependencies (there's a maven equivalent)

I think using the regular JDBC driver dependency should be ok...you just can't mix hibernate reactive/non-reactive...

**Comment by yesvike** [comment]
@kenyee 

That's where I have the problem. Here is my build.gradle

```
dependencies {
    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
    implementation 'io.quarkus:quarkus-hibernate-reactive'
    implementation 'io.quarkus:quarkus-jdbc-db2'
    implementation 'io.quarkus:quarkus-agroal'
    implementation 'io.quarkus:quarkus-reactive-pg-client'
    implementation 'io.quarkus:quarkus-arc'
    implementation 'io.quarkus:quarkus-resteasy'
    testImplementation 'io.quarkus:quarkus-junit5'
    testImplementation 'io.rest-assured:rest-assured'
}
```
application.properties

```
quarkus.datasource.db-kind=postgresql
quarkus.datasource.jdbc=false
quarkus.datasource.reactive.url=****
quarkus.datasource.username=***
quarkus.datasource.password=***


quarkus.datasource.DB2.db-kind=db2
quarkus.datasource.DB2.jdbc.url=***
quarkus.datasource.DB2.username=***
quarkus.datasource.DB2.password=***
```

With this Hibernate reactive ORM is not able to pick the default data source and throws error

```
Exception: Model classes are defined for the default persistence unit, but no default datasource was found. The default EntityManagerFactory will not be created. To solve this, configu
re the default datasource. Refer to https://quarkus.io/guides/datasource for guidance.
        at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.handleHibernateORMWithNoPersistenceXml(HibernateOrmProcessor.java:932)
        at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.configurationDescriptorBuilding(HibernateOrmProcessor.java:420)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:882)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)```

**Comment by filipe-j-delgado** [comment]
I'm getting the same error but in a different use case:
I'm using reactive hibernate with panache and I now also need to add a JDBC config store using Quarkus JDBC Config (https://quarkiverse.github.io/quarkiverse-docs/quarkus-config-extensions/dev/jdbc.html)

`Caused by: java.lang.IllegalStateException: Booting an Hibernate Reactive serviceregistry on a non-reactive RecordedState!`

How can I fix this? Thanks.

**Comment by kenyee** [comment]
FWIW, I stopped using Hibernate Reactive.  Seems like it's a good idea in isolation, but it causes too many incompatibility quirks...

**Comment by arunk-r** [comment]
disabling default jdbc will help the system to focus on reactive configuration.
Here is sample configuration yml data
`quarkus.liquibase.migrate-at-start=false`
`quarkus.liquibase.change-log: db/changelog/master.xml`
`quarkus.datasource.db-kind=postgresql`
`quarkus.datasource.username=${DATASOURCE_USERNAME:user}`
`quarkus.datasource.password=${DATASOURCE_PASSWORD:pwd}`
`quarkus.datasource.jdbc= false`
`quarkus.datasource.url=jdbc:postgresql://${DATASOURCE_HOST:localhost}:${DATASOURCE_PORT:5432}/${DATASOURCE_SCHEMA:database_schema}`
`quarkus.datasource.validation-query-sql=SELECT 1`
`quarkus.datasource.reactive.url=vertx-reactive:postgresql://${DATASOURCE_HOST:localhost}:${DATASOURCE_PORT:5432}/${DATASOURCE_SCHEMA:database_schema}`

**Comment by LazaroR94** [comment]
Has there been any traction on this yet? I've recently got a requirement to use Quartz in my reactive application and need the JDBC jobstore as this will be deployed on a clustered environment.

**Comment by spieps** [comment]
I'm also curious about this. Should we be expecting at some point down the road, to be able to successfully mix reactive and jdbc datasources? 

Right now we can not use `quarkus-quartz` or `quarkus-liquibase` in the same application as `quarkus-hibernate-reactive-panache`

**Comment by Sanne** [comment]
> Should we be expecting at some point down the road, to be able to successfully mix reactive and jdbc datasources?

Eventually yes of course.
In this area unfortunately changes need to be done in this order:
 - move Quarkus to Jakarta EE (lots of work!), so that we can use Hibernate ORM 6
 - Upgrade to Hibernate ORM 6  (lots of work!)
 - Rewrite Hibernate Reactive to be based on Hibernate ORM 6 (..!)
 - Rewrite the Hibernate Reactive / Quarkus extension as 90% of it will need to be different

For these reasons I'm not working on the current Hibernate Reactive extension. N.B. the above tasks are big but they are mostly complete now, except the last one - so I don't expect this will take ages yet.

**Comment by hakuseki** [comment]
This problem applies to Camel as well, when including components dependent on JDBC such as SQL and JDBC. Quartz might be a candidate as well, but that is not verified.
There's no way to include a JDBC datasource at the same time as a reactive one. You'll end up in a Catch 22 with the datasource requirement in Camel vs the above mentioned error.

**Comment by Sanne** [comment]
> There's no way to include a JDBC datasource at the same time as a reactive one. 

I'm not aware of that being a limitation - as long as you don't expect to use both flavours of Hibernate you should be able to define as many datasources as you like, and of different types.
I'd suggest please ask on zulip or open a separate issue.

**Comment by hakuseki** [comment]
Well my tests shows that I can't do it. Maybe it could be that the DevServices is doing something?
I have requested assistance from the Camel User Group and there we can see that if the DevServices is not used, database is already up and running, it works fine.

**Comment by spieps** [comment]
@Sanne @yrodiere  Is mixing reactive and JDBC resources on the roadmap now that Quarkus 3 has released?

**Comment by benjes** [comment]
Is this still an open issue in Quarkus 3? So an app cannot have a reactive and non-reactive datasource? 
I am trying to use the same DB for both reactive and non-reactive:
`  datasource:
    db-kind: mysql
    username: ${DB_USER}
    password: ${DB_PWD}
    reactive:
      url: ${DB_URL}
    jdbc:
      url: jdbc:${DB_URL}`
      
But when trying to use a orm panache repository I get the error:
java.lang.IllegalStateException: The named datasource 'default-reactive' has not been properly configured. See https://quarkus.io/guides/datasource#multiple-datasources for information on how to do that.

The documentation states that a datasource can be both  reactive and non-reactive.

**Comment by yrodiere** [comment]
> Is this still an open issue in Quarkus 3?

Yes, as you can see.

> So an app cannot have a reactive and non-reactive datasource?

An app cannot use both Hibernate ORM and Hibernate Reactive together at the moment.

An app _can_ have a reactive and a non-reactive datasource, provided you only use Hibernate ORM, or only Hibernate Reactive, or neither. And this limitation will stay until this issue gets fixed.

>  ```
> java.lang.IllegalStateException: The named datasource 'default-reactive' has not been properly configured. See https://quarkus.io/guides/datasource#multiple-datasources for information on how to do that.
> ```

I'd need the full stacktrace to say for sure, but from what I can this exception is thrown by `quarkus-hibernate-orm-panache`, whereas the name of your persistence unit suggests you're using with Hibernate Reactive. This won't work, as explained above.

If you want to use Hibernate Reactive, replace your dependency to `quarkus-hibernate-orm-panache` with `quarkus-hibernate-reactive-panache`.
If you want to use Hibernate ORM, remove your dependency to `quarkus-hibernate-reactive`.
If you want to use both, that's not possible until this issue gets fixed.

**Comment by sdlzhd** [comment]
Any progress on this issue?

**Comment by yrodiere** [comment]
> > Is this still an open issue in Quarkus 3?
> 
> Yes, as you can see.
